core/cmd/local_client.go

```diff
@@ -128,19 +129,52 @@ func (cli *Client) runNode(c *clipkg.Context) error {
 		lggr.Warn("Ethereum is disabled. Chainlink will only run services that can operate without an ethereum connection")
 	}
 
+	// DB conn is opened before any services have started,
+	// and closed after all services have stopped.
 	db, err := openDB(cli.Config, lggr)
 	if err != nil {
 		return cli.errorOut(errors.Wrap(err, "opening db"))
 	}
 	defer lggr.ErrorIfClosing(db, "db")
 
-	sig := shutdown.NewSignal()
-	cleanup, err := applicationLockDB(cli.Config, db, lggr, sig)
-	defer cleanup()
+	// rootCtx will be cancelled when SIGINT|SIGTERM is received.
+	rootCtx, cancelRootCtx := context.WithCancel(context.Background())
+
+	// cleanExit is used to skip "fail fast" routine
+	cleanExit := make(chan struct{})
+	var shutdownStartTime time.Time
+	defer func() {
+		close(cleanExit)
+		log.Printf("Graceful shutdown time: %s", time.Since(shutdownStartTime))
+	}()
+
+	go shutdown.HandleShutdown(func() {
+		shutdownStartTime = time.Now()
+		cancelRootCtx()
+
+		select {
+		case <-cleanExit:
+			return
+		case <-time.After(cli.Config.ShutdownGracePeriod()):
+		}
+
+		lggr.Criticalf("Shutdown grace period of %v exceeded, closing DB and exiting...", cli.Config.ShutdownGracePeriod())
+		if err = db.Close(); err != nil {
+			lggr.Criticalf("Failed to close DB: %v", err)
+		}
+		if err = lggr.Sync(); err != nil {
+			log.Printf("Failed to sync Logger: %v", err)
+		}
+
+		os.Exit(-1)
+	})
+
+	waitLocksReleased, err := applicationLockDB(rootCtx, cli.Config, db, lggr)
```

Comment:
This used to defer the cleanup right away, so it was executed in case of error (which is still necessary), but now it could leave something hanging. Either the error check below should run `waitLocksReleased()` (and it must not be nil), or `applicationLockDB` should handle the cleanup in the case of returning an error. I'm thinking the latter would be cleaner, and might simplify away some of the existing code that constructs the func to return.

Reply:
Excellent finding! I will review this once again and change it as you suggested.

Reply:
Nothing gets past @jmank88. Guy has god-like review skills